lightningd/plugin.c

```diff
@@ -638,6 +638,11 @@ static const char *plugin_read_json_one(struct plugin *plugin,
 		return NULL;
 	}
 
+	if (plugin->toks->type != JSMN_OBJECT)
```

Comment:
I am no expert, but isn't this already checked in `json_parse_input` ?

https://github.com/ElementsProject/lightning/blob/b618d662bcb9b7040a2df6d6582fa9fd35073db8/lightningd/plugin.c#L618-L624

Reply:
Yes! but the message is wrong! 

We fail to parse the JSON but messages that are not JSON at all and this makes the debug very bad and difficult, as mentioned in the following sentences

>Having a better error message can speed up the debugging process, and from our point of view it is just an if

Reply:
Ok I see it now,  `json_parse_input` doesn't catch that error.

Tested with a build-in plugin on the command line (which uses the same [json_parse_input](https://github.com/ElementsProject/lightning/blob/f67ab2a86e5d4543570925d9efafcb2b56a6c8d3/plugins/libplugin.c#L704-L707) and I enter your line:
```
./autoclean 
"{valid json rpc 2.0 notification}"
Malformed JSON-RPC notification missing "method" or "params": "{valid json rpc 2.0 notification}"
```
That error is from a later step, passed  `json_parse_input`.

To test manually, maybe modify [cowsay.sh](https://github.com/ElementsProject/lightning/blob/master/contrib/plugins/cowsay.sh)?

Reply:
If you put this line in the command line `"{valid JSON rpc 2.0 notification}"` you are passing a string to the stdio?

BTW the body of the json need to contains a valid json, in this case `\"{jsonrpc: \"2.0\", id: \"1\", \"method\": \"getinfo\", \"params\": \"\"}\"`

The `json_parse_input` parse ok all the JSON received because the string is a valid JSON, and if you repeat the test your test should be ok (my string example may contain typo)

Here the choice to put the check-in there is to make a minimal change to the repository because this is to increase only a debug verbosity to avoid some common mistakes during a plugin API development, and I think that putting this change inside a json_parse_input is a wrong place? maybe we need to have some sanity check file regarding the JSON RPC protocol?

BTW, the string `{any plain content` is the wrong JSON format, and the JSON library is right!

Reply:
> you are passing a string to the stdio

Yep just copy paste the line into the terminal, also handy for debugging plugins separately.
Forget about the test, `cowsay.sh` was just a loose suggestion (but its an insightful example).

Please carry on.
Ack  90f6df5